src/v/cluster/partition_balancer_planner.cc

```diff
@@ -390,31 +367,90 @@ void partition_balancer_planner::get_full_node_reassignments(
               ntp_size_it->first,
-              true,
+              _config.soft_max_disk_usage_ratio,
               rrs);
-            if (new_allocation_units) {
-                result.reassignments.emplace_back(ntp_reassignments{
-                  .ntp = partition_to_move,
-                  .allocation_units = std::move(new_allocation_units.value())});
-            } else {
+
+            struct full_node_replica {
+                model::broker_shard bs;
+                node_disk_space disk;
+            };
+            std::vector<full_node_replica> full_node_replicas;
+            std::vector<model::broker_shard> stable_replicas;
+
+            for (const auto& r : current_assignments->replicas) {
+                if (rrs.unavailable_nodes.contains(r.node_id)) {
+                    continue;
+                }
+
+                auto disk_it = rrs.node_disk_reports.find(r.node_id);
+                vassert(
+                  disk_it != rrs.node_disk_reports.end(),
+                  "disk report for node {} must be present",
+                  r.node_id);
+                const auto& disk = disk_it->second;
+
+                if (
+                  disk.final_used_ratio() < _config.soft_max_disk_usage_ratio) {
+                    stable_replicas.push_back(r);
+                } else {
+                    full_node_replicas.push_back(full_node_replica{
+                      .bs = r,
+                      .disk = disk,
+                    });
+                }
+            }
+
+            // We start with a small set of stable replicas that are on "good"
+            // nodes and try to find a reallocation. If that fails, we add one
+            // replica from the set of full_node_replicas (starting from the
+            // least full) to stable replicas and retry until we get a valid
+            // reallocation.
+            std::sort(
+              full_node_replicas.begin(),
+              full_node_replicas.end(),
+              [](const auto& lhs, const auto& rhs) {
+                  return lhs.disk.final_used_ratio()
+                         < rhs.disk.final_used_ratio();
+              });
+
+            bool success = false;
+            for (const auto& replica : full_node_replicas) {
+                auto new_allocation_units = get_reallocation(
+                  partition_to_move,
+                  *current_assignments,
+                  ntp_size_it->first,
+                  constraints,
+                  stable_replicas,
+                  rrs);
+
+                if (new_allocation_units) {
+                    result.reassignments.emplace_back(ntp_reassignments{
+                      .ntp = partition_to_move,
+                      .allocation_units = std::move(
+                        new_allocation_units.value())});
+                    success = true;
+                    break;
+                } else {
+                    stable_replicas.push_back(replica.bs);
```

Comment:
shold we check here if replica that we are going to add to stable replicas has disk usage ratio smaller than hard limit ?

Reply:
I'm not sure, why we would want that? Stable replicas will remain where they are, so in any case we are not making the situation worse for these nodes. And partial move can make situation better and unlock further moves.

Example: suppose n1, n2, n3 are full (maybe even over hard limit) and n4 is empty. For the ntp with a replica set of 1, 2, 3 we would initially have empty `stable_replicas`. But this move is impossible. So after a couple of retries we will add two nodes with least filled disk (say n2 and n3) to stable replicas and move a replica from n1 to n4, improving the balance.

Reply:
thanks for explaining this